### ISSUE TITLE
There Are No Interface Routes for Non-default Ethernet / Wi-Fi Services

### ISSUE THREAD

**Issue by gerickson** [Nuovations/connman#5]
I am attempting to keep Ethernet + Wi-Fi + Cellular concurrent at all times, in that priority order. I expect, of course, that the most preferred and connected technology will keep the default route. However, I would expect that if I bind to an interface, I can still get Internet reachability with that interface. This works for Ethernet (expected, since it has the default route) and Cellular and Wi-Fi for IPv6. However, Wi-Fi over IPv4 can only reach its local subnet but not the Internet when it is not primary:

I suspect an issue with the routing setup:

```
# ifconfig 
eth0      Link encap:Ethernet  HWaddr 00:04:F3:1A:5E:E0  
          inet addr:192.168.2.69  Bcast:192.168.2.255  Mask:255.255.255.0
          inet6 addr: 2601:647:5a83:8c52:204:f3ff:fe1a:5ee0/64 Scope:Global
          inet6 addr: fe80::204:f3ff:fe1a:5ee0/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:53 errors:0 dropped:0 overruns:0 frame:0
          TX packets:112 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:8488 (8.2 KiB)  TX bytes:16201 (15.8 KiB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:254 errors:0 dropped:0 overruns:0 frame:0
          TX packets:254 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 
          RX bytes:16495 (16.1 KiB)  TX bytes:16495 (16.1 KiB)

wlan0     Link encap:Ethernet  HWaddr 00:04:F3:1A:5E:E1  
          inet addr:192.168.2.46  Bcast:192.168.2.255  Mask:255.255.255.0
          inet6 addr: 2601:647:5a83:8c52:204:f3ff:fe1a:5ee1/64 Scope:Global
          inet6 addr: fe80::204:f3ff:fe1a:5ee1/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:41 errors:0 dropped:0 overruns:0 frame:0
          TX packets:143 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:7538 (7.3 KiB)  TX bytes:15891 (15.5 KiB)

wwan0     Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  
          inet addr:100.87.218.182  P-t-P:100.87.218.182  Mask:255.255.255.252
          UP POINTOPOINT RUNNING NOARP MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:1 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:48 (48.0 B)

# ping -c 4 -I eth0 [dns.google.com](http://dns.google.com/) 
PING [dns.google.com](http://dns.google.com/) (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: seq=0 ttl=116 time=8.249 ms
64 bytes from 8.8.8.8: seq=1 ttl=116 time=9.280 ms
64 bytes from 8.8.8.8: seq=2 ttl=116 time=8.085 ms
64 bytes from 8.8.8.8: seq=3 ttl=116 time=8.804 ms

--- [dns.google.com](http://dns.google.com/) ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 8.085/8.604/9.280 ms

# ping6 -c 4 -I eth0 [dns.google.com](http://dns.google.com/) 
PING [dns.google.com](http://dns.google.com/) (2001:4860:4860::8844): 56 data bytes
64 bytes from 2001:4860:4860::8844: seq=0 ttl=57 time=9.434 ms
64 bytes from 2001:4860:4860::8844: seq=1 ttl=57 time=9.235 ms
64 bytes from 2001:4860:4860::8844: seq=2 ttl=57 time=9.297 ms
64 bytes from 2001:4860:4860::8844: seq=3 ttl=57 time=9.166 ms

--- [dns.google.com](http://dns.google.com/) ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 9.166/9.283/9.434 ms

# ping -c 4 -I wlan0 [dns.google.com](http://dns.google.com/) 
PING [dns.google.com](http://dns.google.com/) (8.8.8.8): 56 data bytes

--- [dns.google.com](http://dns.google.com/) ping statistics ---
4 packets transmitted, 0 packets received, 100% packet loss

# ping6 -c 4 -I wlan0 [dns.google.com](http://dns.google.com/) 
PING [dns.google.com](http://dns.google.com/) (2001:4860:4860::8844): 56 data bytes
64 bytes from 2001:4860:4860::8844: seq=0 ttl=57 time=100.924 ms
64 bytes from 2001:4860:4860::8844: seq=1 ttl=57 time=124.037 ms
64 bytes from 2001:4860:4860::8844: seq=2 ttl=57 time=45.249 ms
64 bytes from 2001:4860:4860::8844: seq=3 ttl=57 time=69.188 ms

--- [dns.google.com](http://dns.google.com/) ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 45.249/84.849/124.037 ms

# ping -c 4 -I wwan0 [dns.google.com](http://dns.google.com/) 
PING [dns.google.com](http://dns.google.com/) (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: seq=0 ttl=113 time=605.749 ms
64 bytes from 8.8.8.8: seq=1 ttl=113 time=287.500 ms
64 bytes from 8.8.8.8: seq=2 ttl=113 time=98.122 ms
64 bytes from 8.8.8.8: seq=3 ttl=113 time=293.822 ms

--- [dns.google.com](http://dns.google.com/) ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 98.122/321.298/605.749 ms

# ip -4 route show
default via 192.168.2.1 dev eth0 
100.87.218.180/30 dev wwan0 proto kernel scope link src 100.87.218.182 
100.87.218.181 dev wwan0 scope link 
192.168.2.0/24 dev wlan0 proto kernel scope link src 192.168.2.46 
192.168.2.0/24 dev eth0 proto kernel scope link src 192.168.2.69 
192.168.2.1 dev eth0 scope link 
192.168.2.1 dev wlan0 scope link 
198.224.173.135 via 100.87.218.181 dev wwan0 
198.224.174.135 via 100.87.218.181 dev wwan0 

# ip -6 route show
2601:647:5a83:8c52::/64 dev wlan0 proto kernel metric 256 expires 2591856sec pref medium
2601:647:5a83:8c52::/64 dev eth0 proto kernel metric 256 expires 2591856sec pref medium
fe80::/64 dev wlan0 proto kernel metric 256 pref medium
fe80::/64 dev eth0 proto kernel metric 256 pref medium
ff00::/8 dev wlan0 metric 256 pref medium
ff00::/8 dev eth0 metric 256 pref medium
default via fe80::f29f:c2ff:fe10:271e dev wlan0 proto ra metric 1024 expires 1656sec hoplimit 64 pref medium
default via fe80::f29f:c2ff:fe10:271e dev eth0 proto ra metric 1024 expires 1656sec hoplimit 64 pref medium
```

@denkenz subsequently suggested the [following via the connman mailing list](https://lore.kernel.org/connman/f6f0a90e-28d6-3a23-9a78-ef5082d9186e@gmail.com/T/#m42bc998bfedbdeaa4df769d835a94c2991fd95d2):

```
ip route add default dev wlan0 table 1234
```

which did not quite work; however, I found the following to work:

```
ip route add default dev wlan0 via 192.168.2.1 metric 1024
```

where 192.168.2.1 is the next hop gateway.

**Comment by gerickson** [comment]
As the title implies, it turns out that this exact same problem bedevils Ethernet as well.

In fact, after doing further development on #6, this blocks successful recovery of Ethernet (or Wi-Fi) on IPv4 because of this issue.

**Comment by gerickson** [comment]
The functions that it would appear that need to be added are:

```c
static int connman_inet_modify_interface_default_route(int cmd, int index,
				const char *gateway)
{
    static const uint32_t table_id = RT_TABLE_MAIN;
    static const uint32_t metric = UINT32_MAX;
    static const unsigned char prefixlen = 0;
    g_autofree char *interface = NULL;

    interface = connman_inet_ifname(index);

    DBG("cmd %d (%s) index %d (%s) gateway %s",
		cmd == RTM_NEWROUTE ?
			"add" :
			cmd == RTM_DELROUTE ?
				"del" :
				"<unknown>",
		index, interface, gateway, metric);

    return iproute_default_modify(cmd, table_id, metric, index, gateway,
		prefixlen);
}

int connman_inet_add_interface_default_route(int index, const char *gateway)
{
    static const int cmd = RTM_NEWROUTE;

    return connman_inet_modify_interface_default_route(cmd, index, gateway);
}

int connman_inet_del_interface_default_route(int index, const char *gateway)
{
    static const int cmd = RTM_DELROUTE;

    return connman_inet_modify_interface_default_route(cmd, index, gateway);
}
```

in _inet.c_ which rely on adding a `metric` argument to the existing `iproute_default_modify`.

Then, it would appear that these new functions need to be instrumented into:

* `set_default_gateway`
* `unset_default_gateway`

in _connection.c_. Basically, when the default gateway is set, then `connman_inet_del_interface_default_route` is invoked to remove the conflicting default interface route and when the default gateway is unset, then `connman_inet_add_interface_default_route` is invoked to add a backfilling default interface route.

**Comment by gerickson** [comment]
It turns out that the `redirect-gateway def1` _trick_ employed by OpenVPN also works, albeit it requires two routing table entries to the `RTA_PRIORITY` _metric_ approach's one entry:

```
ip route add 0.0.0.0/1 dev wlan0 via <gateway>
ip route add 128.0.0.0/1 dev wlan0 via <gateway>
```

both routes are more specific than:

```
ip route add default dev eth0 via <gateway>
```

and are thus chosen when doing:

```
ping -c4 -Iwlan0 dns.google.com
```

**Comment by gerickson** [comment]
Per the attached, at this point, the following test is stable and robust:

* Ethernet as default.
* Wi-Fi as secondary.

Both start out correctly with Ethernet taking the high-priority (that is, metric 0) default route and Wi-Fi taking the low-priority (metric 4294967295) default route.

The DUT Ethernet cable is then removed causing Ethernet link to be lost at the DUT. Wi-Fi correctly drops the low-priority default route and acquires the high-priority default route.

The DUT Ethernet cable is then reconnected causing Ethernet link to be reacquired at the DUT. Wi-Fi correctly drops the low-priority route and Ethernet acquires the high-priority default route and Wi-Fi reacquires the low-priority default route.

The [attached log](https://github.com/Nuovations/connman/files/13573194/connman-github-issue-5-log-output-20231205T142830.txt) details the progress.

**Comment by gerickson** [comment]
On to testing Ethernet + Cellular and Ethernet + Wi-Fi + Cellular.

**Comment by gerickson** [comment]
Ethernet + Cellular works.

**Comment by gerickson** [comment]
Ethernet + Wi-Fi + Cellular does not presently work; however, that is because there is only one low-priority metric, `UINT32_MAX`. I need to have a per network interface or network service low priority metric. @denkenz intimated that _iwd_ uses a metric base + the network interface index since it's a static value.

That could work for connman for a first pass. However, it seems like the "rule of least astonishment" for connman would be that the metric is tied to the service order / priority: lower priority order services have a lower priority (larger) metric.

**Comment by gerickson** [comment]
With the low priority metric modulation in place (using `metric = UINT32_MAX >> interface_index` as a first-pass, proof of concept), Ethernet + Wi-Fi + Cellular works. Starting with all three connected:

```
eth0      Link encap:Ethernet  HWaddr 00:04:F3:1A:5E:E0  
          inet addr:192.168.2.69  Bcast:192.168.2.255  Mask:255.255.255.0
          inet6 addr: 2601:647:5a00:15c2:204:f3ff:fe1a:5ee0/64 Scope:Global
          inet6 addr: fe80::204:f3ff:fe1a:5ee0/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:7926 errors:0 dropped:0 overruns:0 frame:0
          TX packets:4669 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:8396829 (8.0 MiB)  TX bytes:485847 (474.4 KiB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:8379 errors:0 dropped:0 overruns:0 frame:0
          TX packets:8379 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 
          RX bytes:465267 (454.3 KiB)  TX bytes:465267 (454.3 KiB)

wlan0     Link encap:Ethernet  HWaddr 00:04:F3:1A:5E:E1  
          inet addr:192.168.2.46  Bcast:192.168.2.255  Mask:255.255.255.0
          inet6 addr: fe80::204:f3ff:fe1a:5ee1/64 Scope:Link
          inet6 addr: 2601:647:5a00:15c2:204:f3ff:fe1a:5ee1/64 Scope:Global
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:323 errors:0 dropped:0 overruns:0 frame:0
          TX packets:353 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:61915 (60.4 KiB)  TX bytes:35073 (34.2 KiB)

wwan0     Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  
          inet addr:100.115.167.141  P-t-P:100.115.167.141  Mask:255.255.255.252
          UP POINTOPOINT RUNNING NOARP MULTICAST  MTU:1500  Metric:1
          RX packets:41 errors:0 dropped:0 overruns:0 frame:0
          TX packets:68 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:3624 (3.5 KiB)  TX bytes:4996 (4.8 KiB)

default via 192.168.2.1 dev eth0 
default via 100.115.167.142 dev wwan0 metric 16777215 
default via 192.168.2.1 dev wlan0 metric 33554431 
100.115.167.140/30 dev wwan0 proto kernel scope link src 100.115.167.141 
100.115.167.142 dev wwan0 scope link 
192.168.2.0/24 dev eth0 proto kernel scope link src 192.168.2.69 
192.168.2.0/24 dev wlan0 proto kernel scope link src 192.168.2.46 
192.168.2.1 dev wlan0 scope link 
192.168.2.1 dev eth0 scope link 
198.224.173.135 via 100.115.167.142 dev wwan0 
198.224.174.135 via 100.115.167.142 dev wwan0 
PING dns.google.com (8.8.4.4): 56 data bytes
64 bytes from 8.8.4.4: seq=0 ttl=57 time=1035.274 ms

--- dns.google.com ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 1035.274/1035.274/1035.274 ms
PING dns.google.com (2001:4860:4860::8888): 56 data bytes
64 bytes from 2001:4860:4860::8888: seq=0 ttl=116 time=89.272 ms

--- dns.google.com ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 89.272/89.272/89.272 ms
```

I then pull the Ethernet cable at the DUT and Wi-Fi promotes to the default and Cellular retains a low-priority default route as desired:

```
eth0      Link encap:Ethernet  HWaddr 00:04:F3:1A:5E:E0  
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:7966 errors:0 dropped:0 overruns:0 frame:0
          TX packets:4698 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:8400475 (8.0 MiB)  TX bytes:491814 (480.2 KiB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:8543 errors:0 dropped:0 overruns:0 frame:0
          TX packets:8543 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 
          RX bytes:474682 (463.5 KiB)  TX bytes:474682 (463.5 KiB)

wlan0     Link encap:Ethernet  HWaddr 00:04:F3:1A:5E:E1  
          inet addr:192.168.2.46  Bcast:192.168.2.255  Mask:255.255.255.0
          inet6 addr: fe80::204:f3ff:fe1a:5ee1/64 Scope:Link
          inet6 addr: 2601:647:5a00:15c2:204:f3ff:fe1a:5ee1/64 Scope:Global
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:382 errors:0 dropped:0 overruns:0 frame:0
          TX packets:395 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:72400 (70.7 KiB)  TX bytes:41752 (40.7 KiB)

wwan0     Link encap:UNSPEC  HWaddr 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  
          inet addr:100.115.167.141  P-t-P:100.115.167.141  Mask:255.255.255.252
          UP POINTOPOINT RUNNING NOARP MULTICAST  MTU:1500  Metric:1
          RX packets:41 errors:0 dropped:0 overruns:0 frame:0
          TX packets:68 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:3624 (3.5 KiB)  TX bytes:4996 (4.8 KiB)

default via 192.168.2.1 dev wlan0 
default via 100.115.167.142 dev wwan0 metric 16777215 
100.115.167.140/30 dev wwan0 proto kernel scope link src 100.115.167.141 
100.115.167.142 dev wwan0 scope link 
192.168.2.0/24 dev wlan0 proto kernel scope link src 192.168.2.46 
192.168.2.1 dev wlan0 scope link 
198.224.173.135 via 100.115.167.142 dev wwan0 
198.224.174.135 via 100.115.167.142 dev wwan0 
PING dns.google.com (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: seq=0 ttl=116 time=11.627 ms

--- dns.google.com ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 11.627/11.627/11.627 ms
PING dns.google.com (2001:4860:4860::8844): 56 data bytes
64 bytes from 2001:4860:4860::8844: seq=0 ttl=116 time=11.607 ms

--- dns.google.com ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 11.607/11.607/11.607 ms
```

Success!

**Comment by gerickson** [comment]
The total patch set is about 80 patches.

**Comment by gerickson** [comment]
Patches upstreamed: https://lore.kernel.org/connman/20231206235056.322578-1-gerickson@nuovations.com/T/#t

**Comment by gerickson** [comment]
Closed with #107.
